File: desktop/windows/Nexus.Desktop/Update/DesktopUpdateCacheCleaner.cs
using System.IO;
using Nexus.Desktop.Diagnostics;
using Nexus.Desktop.Sidecar;

namespace Nexus.Desktop.Update;

internal static class DesktopUpdateCacheCleaner
{
    private const string LastCleanupVersionFileName = "last-update-cache-cleanup-version.txt";

    public static async Task ClearStaleCachesIfNeededAsync(
        DesktopAppVersion currentVersion,
        DesktopStartupTimeline startupTimeline)
    {
        string versionMarker = $"{currentVersion.Version}-{currentVersion.BuildNumber}";
        string markerPath = Path.Combine(DesktopPaths.ConfigDirectory, LastCleanupVersionFileName);
        string previousMarker = await ReadMarkerAsync(markerPath);
        if (string.Equals(previousMarker, versionMarker, StringComparison.Ordinal))
        {
            startupTimeline.Mark("update_cache.cleanup_reuse", new Dictionary<string, string>
            {
                ["version"] = versionMarker,
            });
            return;
        }

        string updatesDirectory = Path.Combine(DesktopPaths.CacheDirectory, "updates");
        startupTimeline.Mark("update_cache.cleanup_begin", new Dictionary<string, string>
        {
            ["current"] = versionMarker,
            ["previous"] = previousMarker,
        });

        CleanupResult result;
        try
        {
            result = await Task.Run(() => RemoveEntries(updatesDirectory));
        }
        catch (Exception exception)
        {
            startupTimeline.Mark("update_cache.cleanup_failed", new Dictionary<string, string>
            {
                ["current"] = versionMarker,
                ["error"] = exception.Message,
            });
            return;
        }
        if (result.FailedEntries > 0)
        {
            startupTimeline.Mark("update_cache.cleanup_failed", new Dictionary<string, string>
            {
                ["current"] = versionMarker,
                ["failed_entries"] = result.FailedEntries.ToString(),
                ["removed_entries"] = result.RemovedEntries.ToString(),
            });
            return;
        }

        try
        {
            Directory.CreateDirectory(DesktopPaths.ConfigDirectory);
            await File.WriteAllTextAsync(markerPath, versionMarker);
            startupTimeline.Mark("update_cache.cleanup_finished", new Dictionary<string, string>
            {
                ["current"] = versionMarker,
                ["previous"] = previousMarker,
                ["removed_entries"] = result.RemovedEntries.ToString(),
            });
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            startupTimeline.Mark("update_cache.cleanup_failed", new Dictionary<string, string>
            {
                ["current"] = versionMarker,
                ["error"] = exception.Message,
                ["removed_entries"] = result.RemovedEntries.ToString(),
            });
        }
    }

    private static async Task<string> ReadMarkerAsync(string markerPath)
    {
        try
        {
            if (!File.Exists(markerPath))
            {
                return string.Empty;
            }

            return (await File.ReadAllTextAsync(markerPath)).Trim();
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            return string.Empty;
        }
    }

    private static CleanupResult RemoveEntries(string updatesDirectory)
    {
        if (!Directory.Exists(updatesDirectory))
        {
            return new CleanupResult();
        }

        string[] entries;
        try
        {
            entries = Directory.GetFileSystemEntries(updatesDirectory);
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            return new CleanupResult(FailedEntries: 1);
        }

        int removedEntries = 0;
        int failedEntries = 0;
        foreach (string entry in entries)
        {
            try
            {
                if (Directory.Exists(entry))
                {
                    Directory.Delete(entry, recursive: true);
                    removedEntries++;
                }
                else if (File.Exists(entry))
                {
                    File.Delete(entry);
                    removedEntries++;
                }
            }
            catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
            {
                failedEntries++;
            }
        }

        if (failedEntries == 0 && Directory.Exists(updatesDirectory))
        {
            try
            {
                Directory.Delete(updatesDirectory);
            }
            catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
            {
                failedEntries++;
            }
        }

        return new CleanupResult(removedEntries, failedEntries);
    }

    private readonly record struct CleanupResult(int RemovedEntries = 0, int FailedEntries = 0);
}

File: desktop/macos/Sources/NexusDesktop/Update/DesktopUpdateCacheCleaner.swift
import Foundation

enum DesktopUpdateCacheCleaner {
  private static let lastCleanupVersionKey = "com.leemysw.nexus.desktop.lastUpdateCacheCleanupVersion"

  @MainActor
  static func clearStaleCachesIfNeeded(
    currentVersion: DesktopAppVersion,
    startupTimeline: DesktopStartupTimeline,
    defaults: UserDefaults = .standard
  ) async {
    let versionMarker = "\(currentVersion.version)-\(currentVersion.buildNumber)"
    let previousMarker = defaults.string(forKey: lastCleanupVersionKey)
    guard previousMarker != versionMarker else {
      startupTimeline.mark("update_cache.cleanup_reuse", metadata: [
        "version": versionMarker,
      ])
      return
    }

    let updatesDirectory = DesktopPaths.cacheDirectory
      .appendingPathComponent("updates", isDirectory: true)
    startupTimeline.mark("update_cache.cleanup_begin", metadata: [
      "current": versionMarker,
      "previous": previousMarker ?? "",
    ])

    let result = await Task.detached(priority: .utility) {
      Self.removeEntries(at: updatesDirectory)
    }.value
    guard result.failedEntries == 0 else {
      startupTimeline.mark("update_cache.cleanup_failed", metadata: [
        "current": versionMarker,
        "failed_entries": String(result.failedEntries),
        "removed_entries": String(result.removedEntries),
      ])
      return
    }

    defaults.set(versionMarker, forKey: lastCleanupVersionKey)
    startupTimeline.mark("update_cache.cleanup_finished", metadata: [
      "current": versionMarker,
      "previous": previousMarker ?? "",
      "removed_entries": String(result.removedEntries),
    ])
  }

  private static func removeEntries(at updatesDirectory: URL) -> CleanupResult {
    let fileManager = FileManager.default
    guard fileManager.fileExists(atPath: updatesDirectory.path) else {
      return CleanupResult()
    }

    let entries: [URL]
    do {
      entries = try fileManager.contentsOfDirectory(
        at: updatesDirectory,
        includingPropertiesForKeys: nil,
        options: []
      )
    } catch {
      return CleanupResult(failedEntries: 1)
    }

    var removedEntries = 0
    var failedEntries = 0
    for entry in entries {
      do {
        try fileManager.removeItem(at: entry)
        removedEntries += 1
      } catch {
        failedEntries += 1
      }
    }

    if failedEntries == 0, fileManager.fileExists(atPath: updatesDirectory.path) {
      do {
        try fileManager.removeItem(at: updatesDirectory)
      } catch {
        failedEntries += 1
      }
    }

    return CleanupResult(
      removedEntries: removedEntries,
      failedEntries: failedEntries
    )
  }
}

private extension DesktopUpdateCacheCleaner {
  struct CleanupResult: Sendable {
    let removedEntries: Int
    let failedEntries: Int

    init(removedEntries: Int = 0, failedEntries: Int = 0) {
      self.removedEntries = removedEntries
      self.failedEntries = failedEntries
    }
  }
}
